src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -735,9 +735,9 @@ async Task ITokenCacheInternal.ClearAsync()
             {
                 TokenCacheNotificationArgs args = new TokenCacheNotificationArgs(this, ClientId, null, true, (this as ITokenCacheInternal).IsApplicationCache);
 
-                await (this as ITokenCacheInternal).OnBeforeAccessAsync(args).ConfigureAwait(false);
                 try
                 {
+                    await (this as ITokenCacheInternal).OnBeforeAccessAsync(args).ConfigureAwait(false);
                     await (this as ITokenCacheInternal).OnBeforeWriteAsync(args).ConfigureAwait(false);
```

Comment:
Should this be in try/finally similar to what you have been doing elsewhere?

Reply:
It is, in the finally we need to make sure to call `OnAfterAccess`.

Reply:
how about ensuring both of the onbefore calls?

---
In reply to: [405060863](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1740#discussion_r405060863) [](ancestors = 405060863)

Reply:
I'm not sure. If in `BeforeAccess` the developer throws an exception, does it make sense to still invoke `BeforeWrite` ?

Reply:
In general I don't see the purpose of `BeforeWrite` since we have the `HasCacheChanged` flag on `BeforeAccess`. Just makes things harder to maintain ...

Reply:
I suppose I agree regarding the need for both events. Should we get rid of it? #deprecate, perhaps an opt-in (or possibly not worth considering)? 

My thinking was we should be guaranteeing both to be called. We don't know what folks are using those events for. Perhaps read locks? 

---
In reply to: [405065050](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1740#discussion_r405065050) [](ancestors = 405065050)